### PR TITLE
Work on lots of TODOs

### PR DESCRIPTION
**Pull request by aelred**

### CODE REVIEW

File: grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/sets/RolePlayerFragmentSet.java [hunk 1]
```diff
@@ -90,8 +88,9 @@
      * However, we must still retain the {@link LabelFragmentSet} because it is possible it is selected as a result or
      * referred to elsewhere in the query.
      */
-    static boolean applyRolePlayerRoleOptimisation(Collection<EquivalentFragmentSet> fragmentSets, GraknTx graph) {
-        Iterable<RolePlayerFragmentSet> rolePlayers = EquivalentFragmentSets.fragmentSetOfType(RolePlayerFragmentSet.class, fragmentSets)::iterator;
+    static final FragmentSetOptimisation ROLE_OPTIMISATION = (fragmentSets, tx) -> {
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/sets/RolePlayerFragmentSet.java]: Yay for the interface. Makes it much clearer what it's for 👍

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/rule/InferenceRule.java [hunk 2]
```diff
@@ -291,6 +291,8 @@ public Unifier getUnifier(Atom parentAtom) {
         }
         //case of match all relation atom
         else{
+            // TODO: is this cast always safe?
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/rule/InferenceRule.java]: @kasper-piskorski can you answer this?

File: grakn-graql/src/test/java/ai/grakn/graql/internal/reasoner/inference/GenealogyTest.java [hunk 3]
```diff
@@ -181,7 +181,7 @@ public void testMarriageNotEquals(){
     //@Ignore
```
Comment (fppt) [inline grakn-graql/src/test/java/ai/grakn/graql/internal/reasoner/inference/GenealogyTest.java]: Kill this comment

File: grakn-graql/src/test/java/ai/grakn/graql/internal/reasoner/inference/GenealogyTest.java [hunk 4]
```diff
@@ -293,7 +294,7 @@ public void testWife2(){
 
     @Test
     public void testSiblings() {
```
Comment (fppt) [inline grakn-graql/src/test/java/ai/grakn/graql/internal/reasoner/inference/GenealogyTest.java]: Crappy test names in this class

File: grakn-kb/src/main/java/ai/grakn/factory/TxFactoryAbstract.java [hunk 5]
```diff
@@ -62,7 +63,7 @@
     TxFactoryAbstract(String keyspace, String engineUrl, Properties properties){
         Objects.requireNonNull(keyspace);
 
-        this.keyspace = keyspace.toLowerCase();
+        this.keyspace = keyspace.toLowerCase(Locale.getDefault());
```
Comment (fppt) [inline grakn-kb/src/main/java/ai/grakn/factory/TxFactoryAbstract.java]: Thankfully this is dieing

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/rule/InferenceRule.java (hunk 2 above)
Comment (kasper-piskorski) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/rule/InferenceRule.java]: I think it's not safe :( Instead I should make the addType method Atom specific.

File: grakn-graql/src/test/java/ai/grakn/graql/internal/reasoner/inference/GenealogyTest.java (hunk 4 above)
Comment (kasper-piskorski) [reply grakn-graql/src/test/java/ai/grakn/graql/internal/reasoner/inference/GenealogyTest.java]: strictly use-case-based :p

Comment (kasper-piskorski) [reply grakn-graql/src/test/java/ai/grakn/graql/internal/reasoner/inference/GenealogyTest.java]: actually it's sort of an integration test

**Comment by aelred** [comment]
Thanks @kasper-piskorski !

Re-review please @fppt

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/Atom.java [hunk 6]
```diff
@@ -258,6 +258,15 @@ public IdPredicate getIdPredicate(Var var){
      */
     public Set<Unifier> getPermutationUnifiers(Atom headAtom){ return Collections.singleton(new UnifierImpl());}
 
+    @Override
+    public Atom inferTypes(){ return this; }
+
+    /**
+     * @param type to be added to this relation
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/Atom.java]: `{@link Type} {@link Relationship}`

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationAtom.java [hunk 7]
```diff
@@ -475,7 +466,7 @@ private RelationAtom inferRelationType(Answer sub){
     }
 
     @Override
-    public Atom inferTypes() {
+    public RelationAtom inferTypes() {
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationAtom.java]: Maybe this should be `RelationshipAtom`?

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationAtom.java [hunk 8]
```diff
@@ -475,7 +466,7 @@ private RelationAtom inferRelationType(Answer sub){
...
         return this
                 .inferRelationType(new QueryAnswer())
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationAtom.java]: `inferRelationshipType`

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationAtom.java [hunk 9]
```diff
@@ -475,7 +466,7 @@ private RelationAtom inferRelationType(Answer sub){
...
                 .inferRoleTypes();
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationAtom.java]: `inferRoles`

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationAtom.java (hunk 7 above)
Comment (kasper-piskorski) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationAtom.java]: naaaaah

**Pull request merged**